torchgeo/datasets/satlas.py

```diff
@@ -0,0 +1,700 @@
+        'naip': ('tci', 'ir'),
+        'sentinel1': ('vh', 'vv'),
+        'sentinel2': ('tci', 'b05', 'b06', 'b07', 'b08', 'b11', 'b12'),
+    }
+
+    def __init__(
+        self,
+        root: Path = 'data',
+        split: str = 'train_lowres',
+        images: Iterable[str] = ('sentinel1', 'sentinel2', 'landsat'),
+        labels: Iterable[str] = ('land_cover',),
+        transforms: Callable[[dict[str, Tensor]], dict[str, Tensor]] | None = None,
+        download: bool = False,
+        checksum: bool = False,
+    ) -> None:
+        """Initialize a new SatlasPretrain instance.
+
+        Args:
+            root: Root directory where dataset can be found.
+            split: Metadata split to load.
+            images: List of image products.
+            labels: List of label products.
+            transforms: A function/transform that takes input sample and its target as
+                entry and returns a transformed version.
+            download: If True, download dataset and store it in the root directory.
+            checksum: If True, check the MD5 of the downloaded files (may be slow).
+
+        Raises:
+            AssertionError: If *images* is invalid.
+            DatasetNotFoundError: If dataset is not found and *download* is False.
+        """
+        assert set(images) <= set(self.bands.keys())
+
+        self.root = root
+        self.split = split
+        self.images = images
+        self.labels = labels
+        self.transforms = transforms
+        self.download = download
+        self.checksum = checksum
+
+        self._verify()
+
+        with open(os.path.join(root, 'metadata', f'{split}.json')) as f:
+            self.data = torch.tensor(json.load(f))
+
+    def __len__(self) -> int:
+        """Return the number of images in the dataset.
+
+        Returns:
+            Length of the dataset
+        """
+        return len(self.data)
+
+    def __getitem__(self, index: int) -> dict[str, Tensor]:
+        """Return an index within the dataset.
+
+        Args:
+            index: Index to return.
+
+        Returns:
+            Data and label at that index.
+        """
+        col, row = self.data[index]
+        sample: dict[str, Tensor] = {}
+
+        for image in self.images:
+            sample[f'image_{image}'] = self._load_image(image, col, row)
+
+        for label in self.labels:
+            sample[f'mask_{label}'] = self._load_label(label, col, row)
+
+        if self.transforms is not None:
+            sample = self.transforms(sample)
+
+        return sample
+
+    def _load_image(self, image: str, col: int, row: int) -> Tensor:
+        """Load a single image.
+
+        Args:
+            image: Image product.
+            col: Web Mercator column.
+            row: Web Mercator row.
+
+        Returns:
+            An image tensor.
+        """
+        bands = self.bands[image]
+        fname_glob = os.path.join(self.root, image, '*', bands[0], f'{col}_{row}.png')
```

Comment:
This is one pain point I noticed. It isn't clear to me how to map between entries in `train_lowres.json` to an actual file without relying on globbing (slow). Is there always a unique file for each image/band/col/row pair? Or is it possible to have multiple files from different landsat/sentinel/naip scenes at different times? If the former, it would be really helpful to have a JSON file mapping between col/row (the things in `train_lowres.json`) to the directory where this image is stored.

Reply:
> Or is it possible to have multiple files from different landsat/sentinel/naip scenes at different times?

Looks like this is the case. Not exactly sure how to interpret the `train_lowres.json` file or what a training epoch might look like since there isn't a 1-to-1 mapping between epoch index and image file.

Reply:
The static labels are intended to be paired with an arbitrary image or time series of images from 2022 (for Sentinel-2) or 2018-2021 (for NAIP).

The dynamic labels reference a specific image/timestamp.

On each train step, it:
1. Samples a random task, weighted by the number of tiles at which labels are available for that task. Then it samples a random tile. (https://github.com/allenai/satlas/blob/55bb52fb90a9caaa91bf5fb428e76bfd911a5364/satlas/model/dataset.py#L709)
2. Loads the labels at that tile for that task. (https://github.com/allenai/satlas/blob/55bb52fb90a9caaa91bf5fb428e76bfd911a5364/satlas/model/dataset.py#L797)
3. Identifies the available images that correspond with the loaded labels, based on the validity time range of the label.
4. Samples a random subset of the images to use as inputs (1 for single-image model, 8 for Sentinel-2 multi-image model, 4 for NAIP multi-image model). For dynamic labels, there will only be one matching image though; the backbone applies shared encoder on each image and then temporal max pooling across feature maps; so it can input one image for dynamic labels but multiple images for static labels.

Due to the dataset size, we use mini epochs of a fixed duration (32768 batches), but could also train with epochs where every (task, tile) tuple is sampled. The image subset can be thought of as a data augmentation.

Reply:
How does 3 work? Specifically, given a col/row pair, how do you get a list of times/directories where that col/row can be found? Right now I use globbing but it's incredibly slow.

Reply:
We create a cache mapping from column/row to the list of image IDs containing that tile. This is created by reading good_images_X.json from the metadata.tar, which is a list of (col, row, imageID) tuples (`list[tuple[int, int, str]]`), and then converting it to a `dict[tuple[int, int], list[str]]`. For Sentinel-2 it is called `metadata/good_images_lowres_all.json`.

It needs to be combined with `image_times.json` to make sure the timestamp of the image is appropriate for the label though.

The `good_images_X.json` files are generated by [get_good_images.py](https://github.com/allenai/satlas/blob/main/satlas/cmd/to_dataset/get_good_images.py).

Reply:
Thanks, this is exactly what I was looking for!